### PR TITLE
[Cosmos] Migrate get_attachment operation to pipeline architecture

### PR DESCRIPTION
**Pull request by TheeGreenBee**
Moving get_attachment to pipeline architecture.

Continuing #290 

I am pretty new to rust and wanted to get my hands on something real. Let me know if I've made any mistakes! I have tried to use previous migration examples as much as I can.

### CODE REVIEW

File: sdk/cosmos/src/operations/get_attachment.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            service_version: service_version_from_headers(&headers)?.to_owned(),
+            activity_id: activity_id_from_headers(&headers)?,
+            gateway_version: gateway_version_from_headers(&headers)?.to_owned(),
+            date: date_from_headers(&headers)?,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {}
```
Comment (rylev) [inline sdk/cosmos/src/operations/get_attachment.rs:111]: If you don't have any tests, you should remove this.

**Comment by TheeGreenBee** [comment]
@rylev I have been trying to run the test you have pointed me towards locally but I can't get it to work. I have tried:

` $ TESTING_MODE=RECORD cargo run --example create_database_and_collection --features=mock_transport_framework`

which gave this error:

`error: no example target named "create_database_and_collection"`

and 

`cargo test --features mock_transport_framework`

with `TESTING_MODE` set to `RECORD` in my environment. This gives the following error:

```
[2021-10-09T11:49:54Z ERROR azure_core::policies::retry_policies::exponential_retry] Error occurred when making request: failed to execute request: error sending request for url (https://.documents.azure.com/dbs): error trying to connect: dns error: failed to lookup address information: Name or service not known
```

I think the first error is expected, since you moved the create_database_and_collection test out of the examples directory. The second is more interesting. The URL (https://.documents.azure.com/dbs) doesn't contain the cosmos account name. Following the code made me think this issue is arising in `setup.rs`. I can see that tests that don't use the mock_transport_framework get the cosmos account name from an environment variable, `COSMOS_ACCOUNT`, but I can't see how the mock_transport_framework tests get the cosmos account name. 

It would be great for me to be able to run this locally so I can look at the output of a successful RECORD test. What am I missing/doing wrong here?

Thanks

**Comment by dimbleby** [comment]
Some notes from the latest updates - which pipeline-ify more attachment-related APIs

- slugs were a bit awkward.
  - I'm not sure whether adding the Slug type is right just so that I can add this as a header, or whether I've missed some utility function that would have done the job.
  - slug replacement was previously using the CreateSlugAttachmentResponse, I've chosen to copy that into a ReplaceSlugAttachmentResponse, valuing the consistency more than the duplication

- I have removed the explicit setting of content-length and content-type for requests with bodies. 
  - This is also what the other operations do so I assume there's some magic somewhere that adds these headers, but mentioning it in case I'm wrong and reviewer says "oops, we should be setting those".

- We currently don't have a cosmosdb where this API is enabled... so none of this is yet tested against a real system.  We have a support request open so hopefully we'll be able to get to this soon and add the mock transport tests

File: sdk/cosmos/src/operations/create_reference_attachment.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        #[derive(Debug, Serialize)]
+        struct Request<'r> {
+            pub id: &'r str,
+            #[serde(rename = "contentType")]
+            pub content_type: &'r str,
+            pub media: &'r str,
+        }
+
+        let body = azure_core::to_json(&Request {
+            id: self.attachment_client.attachment_name(),
```
Comment (rylev) [inline sdk/cosmos/src/operations/create_reference_attachment.rs:65]: We shouldn't keep a reference to the attachment_client in this options object. Instead we should take the serialized partition key and attachment name as arguments to this function.

File: sdk/cosmos/src/operations/create_reference_attachment.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            content_type: content_type.into().as_str(),
+            media: media.as_ref(),
+        })?;
+
+        request.set_body(body.into());
+        debug!("request == {:#?}", request);
```
Comment (rylev) [inline sdk/cosmos/src/operations/create_reference_attachment.rs:71]: Please remove this

File: sdk/cosmos/src/operations/create_slug_attachment.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        azure_core::headers::add_optional_header2(&self.content_type, request)?;
+
+        crate::cosmos_entity::add_as_partition_key_header_serialized2(
+            self.attachment_client
+                .document_client()
+                .partition_key_serialized(),
+            request,
+        );
+
+        let slug = Slug::new(self.attachment_client.attachment_name());
```
Comment (rylev) [inline sdk/cosmos/src/operations/create_slug_attachment.rs:60]: Same point as above

File: sdk/cosmos/src/operations/delete_attachment.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        if_match_condition: IfMatchCondition<'b> => Some(if_match_condition),
+    }
+
+    pub fn decorate_request(&self, request: &mut HttpRequest) -> Result<(), crate::Error> {
+        azure_core::headers::add_optional_header2(&self.if_match_condition, request)?;
+        azure_core::headers::add_optional_header2(&self.activity_id, request)?;
+        azure_core::headers::add_optional_header2(&self.consistency_level, request)?;
+
+        crate::cosmos_entity::add_as_partition_key_header_serialized2(
+            self.attachment_client
```
Comment (rylev) [inline sdk/cosmos/src/operations/delete_attachment.rs:41]: Same here

File: sdk/cosmos/src/operations/replace_reference_attachment.rs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+use crate::ResourceQuota;
+use azure_core::headers::{etag_from_headers, session_token_from_headers};
+use azure_core::prelude::*;
+use azure_core::SessionToken;
+use azure_core::{collect_pinned_stream, Request as HttpRequest, Response as HttpResponse};
+use chrono::{DateTime, Utc};
+
+#[derive(Debug, Clone)]
+pub struct ReplaceReferenceAttachmentOptions<'a, 'b> {
+    attachment_client: &'a AttachmentClient,
```
Comment (rylev) [inline sdk/cosmos/src/operations/replace_reference_attachment.rs:13]: Same as above

File: sdk/cosmos/src/operations/replace_slug_attachment.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+use crate::{ResourceQuota, Slug};
+use azure_core::headers::{etag_from_headers, session_token_from_headers};
+use azure_core::prelude::*;
+use azure_core::{collect_pinned_stream, Request as HttpRequest, Response as HttpResponse};
+use bytes::Bytes;
+use chrono::{DateTime, Utc};
+
+#[derive(Debug, Clone)]
+pub struct ReplaceSlugAttachmentOptions<'a, 'b> {
+    attachment_client: &'a AttachmentClient,
```
Comment (rylev) [inline sdk/cosmos/src/operations/replace_slug_attachment.rs:12]: Same

File: sdk/cosmos/tests/attachment_00.rs [hunk 8]
```diff
@@ -107,26 +107,35 @@ async fn attachment() -> Result<(), azure_cosmos::Error> {
 
     // create reference attachment
```
Comment (rylev) [inline sdk/cosmos/tests/attachment_00.rs:108]: Can this test be moved to using the mock testing framework?

File: sdk/cosmos/src/operations/create_reference_attachment.rs (hunk 2 above)
Comment (dimbleby) [reply sdk/cosmos/src/operations/create_reference_attachment.rs:65]: I agree with this, but worried that this refactoring was overreach for an MR that was only pipeline-ifying.  However if you're happy for this all to go in one, can take a look later.

File: sdk/cosmos/tests/attachment_00.rs (hunk 8 above)
Comment (dimbleby) [reply sdk/cosmos/tests/attachment_00.rs:108]: Yes, that's the plan - but per https://github.com/Azure/azure-sdk-for-rust/pull/375#issuecomment-945826374 we don't yet have a cosmosdb that exposes this API against which we can run the tests.

**Comment by dimbleby** [comment]
The interesting new commits here are https://github.com/Azure/azure-sdk-for-rust/pull/375/commits/976539c45badefd7782a2104e5fd0dcb0af11a98, https://github.com/Azure/azure-sdk-for-rust/pull/375/commits/6bc2ae2ccbb6c24c3e8fd3a0b67f7191dbd485bd, https://github.com/Azure/azure-sdk-for-rust/pull/375/commits/7bf9b9b8e36620d591cf3d2078344fcf1a94ff0b, https://github.com/Azure/azure-sdk-for-rust/pull/375/commits/41f022b690034dfd5c22c54e1f7de134799a020e, and https://github.com/Azure/azure-sdk-for-rust/pull/375/commits/36525e2aeb5c6eec22b4c8a7e7748a13f57733bd.

- The first two are direct responses to mark-ups from @rylev above
- The third is a simple bug fix 
- The fourth pipeline-ifies `list_attachments()` and `delete_database()`, allowing us to move the attachments testcase over to the mock transport framework
  - The `delete_database()` work was meanwhile also done in main, so that part of this MR is largely erased by subsequent merges
- The fifth makes some fixes arising from running the e2e tests
  - again includes some `delete_database()` work which is undone or duplicated by a later merge

The rest is just merging from main - it's been a busy couple of weeks on main while we've been waiting for everything to come together at our end.

Sorry that this has gotten rather large and unwieldy, I think it is finally ready to go now (modulo any further merges that may be needed...)

File: sdk/cosmos/src/operations/create_slug_attachment.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    activity_id: Option<ActivityId<'a>>,
+    consistency_level: Option<ConsistencyLevel>,
+}
+
+impl<'a> CreateSlugAttachmentOptions<'a> {
+    pub fn new() -> Self {
+        Self {
+            content_type: None,
+            if_match_condition: None,
+            activity_id: None,
```
Comment (rylev) [inline sdk/cosmos/src/operations/create_slug_attachment.rs:25]: I believe activity_id does not need to be set by the user. A pipeline policy will set this on the user's behalf.

File: sdk/cosmos/src/operations/create_slug_attachment.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    pub date: DateTime<Utc>,
+}
+
+impl CreateSlugAttachmentResponse {
+    pub async fn try_from(response: HttpResponse) -> crate::Result<Self> {
+        let (_status_code, headers, pinned_stream) = response.deconstruct();
+        let body = collect_pinned_stream(pinned_stream).await?;
+
+        debug!("headers == {:#?}", headers);
+        debug!("body == {:#?}", body);
```
Comment (rylev) [inline sdk/cosmos/src/operations/create_slug_attachment.rs:98]: Please remove these

File: sdk/cosmos/src/operations/delete_attachment.rs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    if_match_condition: Option<IfMatchCondition<'a>>,
+    activity_id: Option<ActivityId<'a>>,
+    consistency_level: Option<ConsistencyLevel>,
+}
+
+impl<'a> DeleteAttachmentOptions<'a> {
+    pub fn new() -> Self {
+        Self {
+            if_match_condition: None,
+            activity_id: None,
```
Comment (rylev) [inline sdk/cosmos/src/operations/delete_attachment.rs:22]: Same as above

File: sdk/cosmos/src/operations/delete_attachment.rs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    pub gateway_version: String,
+    pub date: DateTime<Utc>,
+}
+
+impl DeleteAttachmentResponse {
+    pub async fn try_from(response: HttpResponse) -> crate::Result<Self> {
+        let (_status_code, headers, pinned_stream) = response.deconstruct();
+        let body = collect_pinned_stream(pinned_stream).await?;
+
+        debug!("headers == {:#?}", &headers);
```
Comment (rylev) [inline sdk/cosmos/src/operations/delete_attachment.rs:81]: Same as above

File: sdk/cosmos/src/operations/list_attachments.rs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    #[serde(rename = "Attachments")]
+    pub attachments: Vec<Attachment>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ListAttachmentsResponse {
+    pub rid: String,
+    pub count: u64,
+    pub attachments: Vec<Attachment>,
+
```
Comment (rylev) [inline sdk/cosmos/src/operations/list_attachments.rs:79]: nit: remove the blank line

File: sdk/cosmos/src/operations/list_attachments.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        max_item_count: i32 => MaxItemCount::new(max_item_count),
+        a_im: ChangeFeed,
+    }
+
+    pub(crate) fn decorate_request(
+        &self,
+        request: &mut HttpRequest,
+        partition_key_serialized: &str,
+    ) -> crate::Result<()> {
+        // add trait headers
```
Comment (rylev) [inline sdk/cosmos/src/operations/list_attachments.rs:47]: This comment can be removed from everywhere

File: sdk/cosmos/src/operations/list_attachments.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+            if_match_condition: None,
+            activity_id: None,
+            consistency_level: None,
+            max_item_count: MaxItemCount::new(-1),
+            a_im: ChangeFeed::None,
+        }
+    }
+
+    setters! {
+        activity_id: &'a str => Some(ActivityId::new(activity_id)),
```
Comment (rylev) [inline sdk/cosmos/src/operations/list_attachments.rs:35]: remove

File: sdk/cosmos/src/operations/list_attachments.rs [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    pub date: DateTime<Utc>,
+    pub continuation_token: Option<String>,
+}
+
+impl ListAttachmentsResponse {
+    pub async fn try_from(response: HttpResponse) -> crate::Result<Self> {
+        let (_status_code, headers, pinned_stream) = response.deconstruct();
+        let body = collect_pinned_stream(pinned_stream).await?;
+
+        debug!("headers == {:#?}", headers);
```
Comment (rylev) [inline sdk/cosmos/src/operations/list_attachments.rs:107]: remove

File: sdk/cosmos/src/operations/replace_slug_attachment.rs [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    activity_id: Option<ActivityId<'a>>,
+    consistency_level: Option<ConsistencyLevel>,
+}
+
+impl<'a> ReplaceSlugAttachmentOptions<'a> {
+    pub fn new() -> Self {
+        Self {
+            content_type: None,
+            if_match_condition: None,
+            activity_id: None,
```
Comment (rylev) [inline sdk/cosmos/src/operations/replace_slug_attachment.rs:23]: remove

**Review by rylev — CHANGES_REQUESTED** [review]
Some small nits but otherwise looks ready to merge. Can you quickly take care of the nits, and then I can hit the merge button?

File: sdk/cosmos/src/operations/replace_slug_attachment.rs [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        azure_core::headers::add_optional_header2(&self.activity_id, request)?;
+        azure_core::headers::add_optional_header2(&self.consistency_level, request)?;
+        azure_core::headers::add_optional_header2(&self.content_type, request)?;
+
+        crate::cosmos_entity::add_as_partition_key_header_serialized2(partition_key, request);
+
+        let slug = Slug::new(attachment_name);
+        azure_core::headers::add_mandatory_header2(&slug, request)?;
+
+        let body = body.into();
```
[the remark this answers is not in the document]
Comment (dimbleby) [reply sdk/cosmos/src/operations/replace_slug_attachment.rs:56]: I think this one is right: the first `into()` gives us `Bytes` and the second gives us a `Body`.

[the remark this answers is not in the document]
Comment (rylev) [reply sdk/cosmos/src/operations/replace_slug_attachment.rs:52]: Hmm I surprised the type inference works for this. It certainly looks funny.

[the remark this answers is not in the document]
Comment (dimbleby) [reply sdk/cosmos/src/operations/replace_slug_attachment.rs:52]: @rylev please say if you're expecting any action on this, else I think this is all good to go

**Comment by rylev** [comment]
I'm really sorry we've let this go for so long. Unfortunately the diff is really large and very hard to review. If you're still interested in getting this merged, I suggest creating new PRs where smaller changes are included in each individual PR. For example, you can see how #584 was closed in favor of smaller more easy to review PRs like #586 and #592. This makes it much easier to review and changes can land much quicker.

**Pull request closed without merging**